website/blog/deployers.md

```diff
@@ -0,0 +1,610 @@
+weavelet. The second runs in the opposite direction from `weaveletWriter` in the
+weavelet to `envelopeReader` in the envelope. We launch the binary in a
+subprocess. The `weaveletReader` and `weaveletWriter` ends of the pipes will be
+accessible in the subprocess as file descriptors 3 and 4 (see [the exec
+package][exec] for details). The weavelet running in the subprocess opens the
+file descriptors stored in the `ENVELOPE_TO_WEAVELET_FD` and
+`WEAVELET_TO_ENVELOPE_FD` environment variables to establish its connection to
+the envelope.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1. Run the binary and establish the pipes between the envelope and
+    // the weavelet.
+    //
+    //              envelope               weavelet
+    //             ┌──────────────────┐   ┌───────────────────┐
+    //             │ envelopeWriter --│---│-> weaveletReader  │
+    //             │ envelopeReader <-│---│-- weaveletWriter  │
+    //             └──────────────────┘   └───────────────────┘
+    weaveletReader, envelopeWriter, err := os.Pipe()
+    if err != nil {
+        return err
+    }
+    envelopeReader, weaveletWriter, err := os.Pipe()
+    if err != nil {
+        return err
+    }
+
+    cmd := exec.Command(binary)
+    cmd.ExtraFiles = []*os.File{weaveletReader, weaveletWriter}
+    cmd.Env = []string{"ENVELOPE_TO_WEAVELET_FD=3", "WEAVELET_TO_ENVELOPE_FD=4"}
+    if err := cmd.Start(); err != nil {
+        return err
+    }
+
+    // ...
+}
+```
+
+Second, the envelope sends an `EnvelopeMsg` containing an `EnvelopeInfo` to the
+weavelet. This is the first message a weavelet expects to receive. An
+`EnvelopeInfo` provides the weavelet with a basic set of metadata including the
+name of the app, the unique id of the deployment, the unique id of the weavelet,
+and so on.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1...
+
+    // Step 2. Send an EnvelopeInfo to the weavelet.
+    info := &protos.EnvelopeMsg{
+        EnvelopeInfo: &protos.EnvelopeInfo{
+            App:           "app",
+            DeploymentId:  uuid.New().String(),
+            Id:            uuid.New().String(),
+            SingleProcess: false,
+            SingleMachine: true,
+            RunMain:       true,
+        },
+    }
+    if err := protomsg.Write(envelopeWriter, info); err != nil {
+        return err
+    }
+
+    // ...
+}
+```
+
+Third, the deployer reads and prints a `WeaveletMsg` from the weavelet.
+
+```go
+func run(ctx context.Context, binary string) error {
+    // Step 1...
+    // Step 2...
+
+    // Step 3. Receive a WeaveletInfo from the weavelet.
+    var reply protos.WeaveletMsg
+    if err := protomsg.Read(envelopeReader, &reply); err != nil {
+        return err
+    }
+    fmt.Println(protojson.Format(&reply))
+
+    // ...
+}
+```
+
+This `WeaveletMsg` contains a `WeaveletInfo` which contains information about
+the weavelet, specifically it's address and PID:
+
+```json
```

Comment:
Since we are sending protos, should we use text format syntax for protos instead of json?  Or maybe a Go literal that creates the proto?